src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/RazorLSPHandlerBase.cs

```diff
@@ -0,0 +1,151 @@
+
+        protected LSPDocumentSynchronizer DocumentSynchronizer { get; }
+
+        protected virtual Task<TOut> RequestServerAsync<TIn, TOut>(
+            ILanguageClientBroker languageClientBroker,
+            string method,
+            LanguageServerKind serverKind,
+            TIn parameters,
+            CancellationToken cancellationToken)
+        {
+            if (languageClientBroker is null)
+            {
+                throw new ArgumentNullException(nameof(languageClientBroker));
+            }
+
+            if (string.IsNullOrEmpty(method))
+            {
+                throw new ArgumentException("message", nameof(method));
+            }
+
+            var contentType = RazorLSPContentTypeDefinition.Name;
+            if (serverKind == LanguageServerKind.CSharp)
+            {
+                contentType = CSharpVirtualDocumentFactory.CSharpLSPContentTypeName;
+            }
+            else if (serverKind == LanguageServerKind.Html)
+            {
+                contentType = HtmlVirtualDocumentFactory.HtmlLSPContentTypeName;
+            }
+
+            return languageClientBroker.RequestAsync<TIn, TOut>(
+                new[] { contentType },
+                cap => true,
+                method,
+                parameters,
+                cancellationToken);
+        }
+
+        protected virtual async Task<ProjectionResult> GetProjectionAsync(LSPDocumentSnapshot documentSnapshot, Position position, CancellationToken cancellationToken)
+        {
+            var languageQueryParams = new RazorLanguageQueryParams()
+            {
+                Position = new OmniSharpPosition(position.Line, position.Character),
+                Uri = documentSnapshot.Uri
+            };
+
+            var languageResponse = await RequestServerAsync<RazorLanguageQueryParams, RazorLanguageQueryResponse>(
+                LanguageClientBroker,
+                LanguageServerConstants.RazorLanguageQueryEndpoint,
+                LanguageServerKind.Razor,
+                languageQueryParams,
+                cancellationToken);
+
+            VirtualDocumentSnapshot virtualDocument;
+            if (languageResponse.Kind == RazorLanguageKind.CSharp &&
+                documentSnapshot.TryGetVirtualDocument<CSharpVirtualDocumentSnapshot>(out var csharpDoc))
+            {
+                virtualDocument = csharpDoc;
+            }
+            else if (languageResponse.Kind == RazorLanguageKind.Html &&
+                documentSnapshot.TryGetVirtualDocument<HtmlVirtualDocumentSnapshot>(out var htmlDoc))
+            {
+                virtualDocument = htmlDoc;
+            }
+            else
+            {
+                return null;
+            }
+
+            if (languageResponse.HostDocumentVersion == UndefinedDocumentVersion)
+            {
+                // There should always be a document version attached to an open document.
+                // TODO: Log it and move on as if it was synchronized.
```

Comment:
How do we log here?

Reply:
You can log to the activity log but it's not super visible. I'd log there and Debug.Fail:

Here's how we log errors, you can log non-errors as well:
https://github.com/dotnet/aspnetcore-tooling/blob/a0f99cd14f2ded43a0087be52e29e67f5aaaba71/src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/VisualStudioErrorReporter.cs#L36-L39